File: java/src/main/java/io/binac/leetcode/WordBreakII.java
package io.binac.leetcode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * <p>Given a string <code>s</code> and a dictionary of strings <code>wordDict</code>, add spaces in <code>s</code> to construct a sentence where each word is a valid dictionary word. Return all such possible sentences in <strong>any order</strong>.</p>
 *
 * <p><strong>Note</strong> that the same word in the dictionary may be reused multiple times in the segmentation.</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
 * <strong>Output:</strong> ["cats and dog","cat sand dog"]
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "pineapplepenapple", wordDict = ["apple","pen","applepen","pine","pineapple"]
 * <strong>Output:</strong> ["pine apple pen apple","pineapple pen apple","pine applepen apple"]
 * <strong>Explanation:</strong> Note that you are allowed to reuse a dictionary word.
 * </pre>
 *
 * <p><strong>Example 3:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
 * <strong>Output:</strong> []
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>1 &lt;= s.length &lt;= 20</code></li>
 * 	<li><code>1 &lt;= wordDict.length &lt;= 1000</code></li>
 * 	<li><code>1 &lt;= wordDict[i].length &lt;= 10</code></li>
 * 	<li><code>s</code> and <code>wordDict[i]</code> consist of only lowercase English letters.</li>
 * 	<li>All the strings of <code>wordDict</code> are <strong>unique</strong>.</li>
 * </ul>
 */
public class WordBreakII {
    public static class Solution1 {
        private void wordBreak(String s, int i, boolean[][] valid, List<String> result, List<String> path) {
            final int len = s.length();
            if (i == len) {
                result.add(String.join(" ", path));
                return;
            }
            for (int j = i + 1; j <= len; ++j) {
                if (!valid[i][j]) continue;
                path.add(s.substring(i, j));
                wordBreak(s, j, valid, result, path);
                path.remove(path.size() - 1);
            }
        }

        public List<String> wordBreak(String s, List<String> wordDict) {
            final int len = s.length();
            Set<String> dict = new HashSet<>(wordDict);
            boolean[] dp = new boolean[len + 1];
            dp[0] = true;
            boolean[][] valid = new boolean[len][len + 1];
            for (int i, j = 1; j <= len; ++j) {
                for (i = j - 1; i >= 0; --i) {
                    if (dp[i] && dict.contains(s.substring(i, j))) {
                        dp[j] = true;
                        valid[i][j] = true;
                    }
                }
            }
            List<String> result = new ArrayList<>();
            if (!dp[len]) return result;

            wordBreak(s, 0, valid, result, new ArrayList<>());
            return result;
        }
    }
}

File: python3/leetcodepy/word_break_ii.py
"""
Given a string s and a dictionary of strings wordDict, add spaces in s to construct a sentence where each word is a valid dictionary word. Return all such possible sentences in any order.

Note that the same word in the dictionary may be reused multiple times in the segmentation.



Example 1:

Input: s = "catsanddog", wordDict = ["cat","cats","and","sand","dog"]
Output: ["cats and dog","cat sand dog"]

Example 2:

Input: s = "pineapplepenapple", wordDict = ["apple","pen","applepen","pine","pineapple"]
Output: ["pine apple pen apple","pineapple pen apple","pine applepen apple"]
Explanation: Note that you are allowed to reuse a dictionary word.

Example 3:

Input: s = "catsandog", wordDict = ["cats","dog","sand","and","cat"]
Output: []


Constraints:

1 <= s.length <= 20
1 <= wordDict.length <= 1000
1 <= wordDict[i].length <= 10
s and wordDict[i] consist of only lowercase English letters.
All the strings of wordDict are unique.
"""
from typing import List


class Solution1:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        size = len(s)
        word_set = set(wordDict)
        dp = [False] * (size + 1)
        dp[0] = True
        valid = [[False] * (size + 1) for _ in range(size)]
        for j in range(1, size + 1):
            for i in range(j - 1, -1, -1):
                if dp[i] and s[i:j] in word_set:
                    dp[j] = True
                    valid[i][j] = True
        result = []
        if not dp[size]:
            return result
        self.__word_break(s, 0, valid, result, [])
        return result

    def __word_break(self, s: str, i: int, valid: List[List[bool]], result: List[str], path: List[str]):
        if i == len(s):
            result.append(' '.join(path))
            return
        for j in range(i + 1, len(s) + 1):
            if valid[i][j]:
                path.append(s[i:j])
                self.__word_break(s, j, valid, result, path)
                path.pop()
